snuba/clickhouse/value_types.py

```diff
@@ -0,0 +1,32 @@
+from dataclasses import dataclass
+from typing import Sequence, Union
+
+# We define some dataclasses here to make it harder to accidentally
+# pass unescaped user strings to ClickHouse in ValuesRowEncoder (though
+# at the time of writing we send no user-generated text input to Clickhouse
+# through ValuesRowEncoder)
+
+
+@dataclass
+class ClickhouseInt:
+    val: int
+
+
+@dataclass
+class ClickhouseNumArray:
+    """
+    An array of either floats or ints
+    """
+
+    val: Union[Sequence[float], Sequence[int]]
+
+
+@dataclass
+class ClickhouseUnguardedExpression:
+    """
+    This is unescaped text that will be sent to Clickhouse and evaluated
+    as an expression so we want to make it clear from naming
+    that using this can be unsafe
+    """
+
+    val: str
```

Comment:
Not sure I agree with this approach. I think the encoder should jsut properly escape what it gets, thus work on any string. No need to care if the string is unsafe.

Reply:
See my comment above. This isn't meant to represent a clickhouse string literal but rather a clickhouse raw expression

Reply:
Why not just using what we already have to format queries?
Create an AST expression:
`FunctionCall(whatever whatever)`
And then, in the encoder, just format it with the formatter we already have
https://github.com/getsentry/snuba/blob/master/snuba/clickhouse/formatter/expression.py#L172

It would take care of any sort of escaping and formatting for you and it is thoroughly tested as we use it to format the AST into a clickhosue query.

The downside is that it could increase the CPU usage, we would have to try. I think it would be fine to do that as a follow up

Reply:
Sure, I'm fine with that